File: 74751_lab4_task4.cpp
#include <iostream>

using namespace std;

double calculate_avg_temperature(const double* voltage_readings, int size) {
    if (size == 0) return 0;
    double sum_temp = 0;
    for (int i = 0; i < size; ++i) {
        sum_temp += voltage_readings[i] * 10.0; // Conversion formula
    }
    return sum_temp / size;
}

int filter_noisy_readings(double* readings, int& size, double valid_min, double valid_max) {
    int valid_count = 0;
    for (int i = 0; i < size; ++i) {
        if (readings[i] >= valid_min && readings[i] <= valid_max) {
            readings[valid_count] = readings[i];
            valid_count++;
        }
    }
    size = valid_count; // update effective size
    return valid_count;
}

int main() {
    double voltages[] = {3.5, 3.6, 1.2, 3.8, 9.9, 3.7};
    int size = sizeof(voltages) / sizeof(voltages[0]);
    
    double avg_temp = calculate_avg_temperature(voltages, size);
    cout << "Average Temperature: " << avg_temp << endl;
    
    int valid_cnt = filter_noisy_readings(voltages, size, 3.0, 4.0);
    cout << "Valid count: " << valid_cnt << ", Filtered readings: ";
    for (int i = 0; i < size; ++i) cout << voltages[i] << " ";
    cout << endl;
    
    return 0;
}

File: 74751_lab4_task4.py
def calculate_avg_temperature(voltage_readings):
    if not voltage_readings:
        return 0
    total_temp = sum(v * 10 for v in voltage_readings) # Temp conversion
    return total_temp / len(voltage_readings)

def filter_noisy_readings(readings, valid_min, valid_max):
    valid_count = 0
    for i in range(len(readings)):
        if valid_min <= readings[i] <= valid_max:
            readings[valid_count] = readings[i]
            valid_count += 1
    # Truncate the list to effective size
    del readings[valid_count:]
    return valid_count

if __name__ == "__main__":
    voltages = [3.5, 3.6, 1.2, 3.8, 9.9, 3.7]
    avg_temp = calculate_avg_temperature(voltages)
    print(f"Average Temperature: {avg_temp:.2f}")
    
    valid_cnt = filter_noisy_readings(voltages, 3.0, 4.0)
    print(f"Valid count: {valid_cnt}, Filtered readings: {voltages}")
